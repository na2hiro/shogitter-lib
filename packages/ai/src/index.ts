export * from "./playerRegistry.js";
export { loadModuleOnServer } from "./players/all.js";
export type * from "./Player.js";
