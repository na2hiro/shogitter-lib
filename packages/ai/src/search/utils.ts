import { ShogitterAiException } from "../ShogitterAiException.js";

export function randomSelect<T>(arr: T[]) {
  if (arr.length === 0)
    throw new ShogitterAiException("randomSelect: empty array");
  return arr[Math.floor(Math.random() * arr.length)];
}
