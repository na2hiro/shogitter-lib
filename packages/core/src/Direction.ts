export enum Direction {
  BLACK,
  WHITE,
  LEFT,
  RIGHT,

  SOMETHING = 9, // TODO what is this?
}
