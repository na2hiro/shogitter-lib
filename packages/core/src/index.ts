import Shogi from "./Shogi.js";
import { Species } from "./Ban.js";
import { PlayerInfo } from "./Teban.js";
import { shogitterDB } from "./ShogitterDB.js";
import XY from "./XY.js";
import { KomaInfo } from "./db/koma.js";
import * as quantumUtils from "./utils/quantumUtils.js";

export * from "./Shogi.js";
export * from "./Command.js";
export * from "./Direction.js";
export * from "./Koma.js";

export { Shogi, Species, KomaInfo, PlayerInfo, shogitterDB, XY, quantumUtils };
