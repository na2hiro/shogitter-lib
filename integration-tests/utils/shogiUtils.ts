import Shogi from "../../src/Shogi";
import XY from "../../src/XY";
import {Species} from "../../src/Ban";
import {serialize} from "./ShogiSerializer";
import {Direction} from "../../src/Direction";

export function move(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    try {
        rawMove(shogi, fromX, fromY, toX, toY, nari);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawMove(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    shogi.runCommand({
        type: "move",
        from: [fromX, fromY],
        to: [toX, toY],
        nari
    })
}

export function put(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    try {
        rawPut(shogi, toX, toY, species, id);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawPut(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    shogi.runCommand({
        type: "put",
        to: [toX, toY],
        put: species,
        direction: shogi.teban.getNowDirection(),
        id
    })
}

export function assertCellIsNull(shogi: Shogi, x: number, y: number) {
    expect(shogi.ban.get(new XY(x, y)).isNull()).toBeTruthy;
}

export function assertCell(shogi: Shogi, x: number, y: number, direction: Direction, species: Species) {
    const koma = shogi.ban.get(new XY(x, y));
    expect(koma.isNull()).toBeFalsy();
    expect(koma.species).toBe(species);
    expect(koma.direction).toBe(direction);
}
