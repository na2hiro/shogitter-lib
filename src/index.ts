import Shogi from "./Shogi";
import {Direction} from "./Direction";
import {Species} from "./Ban";
import { PlayerInfo } from "./Teban";
import { KomaInfo, shogitterDB } from "./ShogitterDB";
import XY from "./XY";

export * from "./Shogi";
export default Shogi;
export {Direction, Species, KomaInfo, PlayerInfo, shogitterDB, XY};