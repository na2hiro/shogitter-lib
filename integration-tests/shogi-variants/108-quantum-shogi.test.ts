import Shogi from "../../src/Shogi";
import { move, put } from "../utils/shogiUtils";
import XY from "../../src/XY";
import { Direction } from "../../src/Direction";

describe("Quantum shogi", () => {
  let shogi: Shogi;
  beforeEach(() => {
    shogi = Shogi.ofRuleId(108);
    shogi.start();
  });
  it("can move", () => {
    move(shogi, 7, 7, 6, 5);
    expect(shogi).toMatchSnapshot();
    expect(shogi.ban.get(new XY(6, 5)).getMovable()).toMatchSnapshot();
  });
  it("can go back and forget", () => {
    move(shogi, 7, 7, 5, 5);
    shogi.runCommand({ type: "rollback" });
    move(shogi, 7, 7, 5, 5);
  });
  it("can go back and forget", () => {
    move(shogi, 6, 7, 5, 5);
    move(shogi, 1, 3, 1, 5);
    move(shogi, 9, 7, 8, 5);
    shogi.runCommand({ type: "rollback" });
    move(shogi, 9, 7, 8, 5);
  });
  it("can go back and forget, including capture", () => {
    move(shogi, 6, 7, 5, 5);
    move(shogi, 5, 3, 5, 5);
    move(shogi, 4, 7, 5, 5);
    shogi.runCommand({ type: "rollback" });
    move(shogi, 4, 7, 5, 5);
  });
  it("can put piece with id", () => {
    move(shogi, 7, 7, 6, 5);
    move(shogi, 6, 3, 6, 5);
    move(shogi, 5, 7, 5, 3, true);
    move(shogi, 4, 3, 5, 3);
    move(shogi, 1, 7, 1, 6);
    put(shogi, 9, 5, "ff", 0);
    expect(shogi.ban.get(new XY(9, 5)).getMovable()).toMatchSnapshot();
    shogi.runCommand({ type: "rollback" });
    put(shogi, 9, 5, "ff", 1);
    expect(shogi.ban.get(new XY(9, 5)).getMovable()).toMatchSnapshot();
  });
  // King is not determined after everything is captured
  it("can finish when everything is taken", () => {
    move(shogi, 9, 7, 9, 3, true);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 9, 3, 8, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 8, 3, 8, 2);
    move(shogi, 6, 1, 6, 2);
    move(shogi, 8, 2, 8, 1);
    move(shogi, 7, 1, 7, 2);
    move(shogi, 8, 1, 7, 2);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 7, 2, 7, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 7, 3, 6, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 6, 3, 5, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 5, 3, 4, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 4, 3, 3, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 3, 3, 2, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 2, 3, 1, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 1, 3, 1, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 1, 1, 2, 2);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 2, 2, 2, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 2, 1, 3, 1);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 3, 1, 4, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 4, 1, 5, 1);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 5, 1, 6, 2);
    expect(shogi).toMatchSnapshot();
    move(shogi, 9, 1, 9, 2);
    expect(shogi).toMatchSnapshot();
    //expect(shogi.ban.get(new XY(9, 2)).getMovable()).toHaveLength(5) // of king
    move(shogi, 6, 2, 9, 2);
    expect(shogi).toMatchSnapshot();
  });
  it("can finish when all other pieces has been captured once", () => {
    move(shogi, 9, 7, 9, 3, true);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 9, 3, 8, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 8, 3, 7, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 7, 3, 6, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 6, 3, 5, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 5, 3, 4, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 4, 3, 3, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 3, 3, 2, 3);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 2, 3, 1, 3);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 1, 3, 2, 2);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 2, 2, 1, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 1, 1, 2, 1);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 2, 1, 3, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 3, 1, 4, 1);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 4, 1, 5, 1);
    move(shogi, 9, 1, 9, 2);
    move(shogi, 5, 1, 6, 1);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 6, 1, 7, 1);
    move(shogi, 9, 1, 9, 2);
    put(shogi, 9, 1, "ff", 0);
    move(shogi, 9, 2, 9, 1);
    move(shogi, 7, 1, 8, 1);
    put(shogi, 7, 1, "ff", 0);
    move(shogi, 8, 1, 9, 1);
    move(shogi, 7, 1, 7, 2);
    expect(shogi).toMatchSnapshot();
    move(shogi, 9, 1, 8, 2);
  });
  it("put 2 pieces in a row", () => {
    move(shogi, 7, 7, 7, 3, true);
    move(shogi, 8, 1, 7, 3, true);
    move(shogi, 7, 9, 7, 3, true);
    move(shogi, 6, 1, 7, 3, true);
    put(shogi, 5, 5, "ff", 0);
    move(shogi, 1, 3, 1, 4);
    put(shogi, 4, 5, "ff", 0);
  });
  it("Resign should contain quantum data", () => {
    move(shogi, 7, 7, 7, 3, true);
    const beforeResign = shogi.getObject();
    shogi.runCommand({
      direction: Direction.BLACK,
      type: "resign",
    });
    // No quantum situation should be changed due to resign
    expect(shogi.kifu.arrayKifu[0].data.quantum).toEqual(
      shogi.kifu.arrayKifu[1].data.quantum
    );

    const shogi2 = Shogi.ofJkf(beforeResign);
    shogi2.runCommand({
      direction: Direction.BLACK,
      type: "resign",
    });
    // No quantum situation should be changed due to resign
    expect(shogi2.kifu.arrayKifu[0].data.quantum).toEqual(
      shogi2.kifu.arrayKifu[1].data.quantum
    );
  });
});
