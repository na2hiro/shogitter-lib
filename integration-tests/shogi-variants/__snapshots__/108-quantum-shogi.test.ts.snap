// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantum shogi can finish when all other pieces has been captured once 1`] = `
ban:
(1) 
0fg________________________
___1ff1ff__________________
___________________________
___________________________
___________________________
___________________________
___0ff0ff0ff0ff0ff0ff0ff0ff
___0ff_______________0ff___
0ff0ff0ff0ff0ff0ff0ff0ff0ff
(0) ff18 
kifu:
☗9三量子成 [0,[9,3,[1,"ff"],[0,"fg"]],[9,7,[0,"ff"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗8三成量子 [0,[8,3,[1,"ff"],[0,"fg"]],[9,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗7三成量子 [0,[7,3,[1,"ff"],[0,"fg"]],[8,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗6三成量子 [0,[6,3,[1,"ff"],[0,"fg"]],[7,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗5三成量子 [0,[5,3,[1,"ff"],[0,"fg"]],[6,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗4三成量子 [0,[4,3,[1,"ff"],[0,"fg"]],[5,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗3三成量子 [0,[3,3,[1,"ff"],[0,"fg"]],[4,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2三成量子 [0,[2,3,[1,"ff"],[0,"fg"]],[3,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗1三成量子 [0,[1,3,[1,"ff"],[0,"fg"]],[2,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2二成量子 [0,[2,2,[1,"ff"],[0,"fg"]],[1,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗1一成量子 [0,[1,1,[1,"ff"],[0,"fg"]],[2,2,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2一成量子 [0,[2,1,[1,"ff"],[0,"fg"]],[1,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗3一成量子 [0,[3,1,[1,"ff"],[0,"fg"]],[2,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗4一成量子 [0,[4,1,[1,"ff"],[0,"fg"]],[3,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗5一成量子 [0,[5,1,[1,"ff"],[0,"fg"]],[4,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗6一成量子 [0,[6,1,[1,"ff"],[0,"fg"]],[5,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗7一成量子 [0,[7,1,[1,"ff"],[0,"fg"]],[6,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗9一量子 [0,[9,1,[],[0,"ff"]],[-1,0,"ff"]]
☖同量子 [1,[9,1,[0,"ff"],[1,"ff"]],[9,2,[1,"ff"],[]],[1,1,"ff"]]
☗8一成量子 [0,[8,1,[1,"ff"],[0,"fg"]],[7,1,[0,"fg"],[]],[1,0,"ff"]]
☖7一量子打 [1,[7,1,[],[1,"ff"]],[-1,1,"ff"]]
☗9一成量子 [0,[9,1,[1,"ff"],[0,"fg"]],[8,1,[0,"fg"],[]],[1,0,"ff"]]
☖7二量子 [1,[7,2,[],[1,"ff"]],[7,1,[1,"ff"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 108,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Quantum shogi can finish when everything is taken 1`] = `
ban:
(1) 
1ff________________________
_________0fg_______________
___________________________
___________________________
___________________________
___________________________
___0ff0ff0ff0ff0ff0ff0ff0ff
___0ff_______________0ff___
0ff0ff0ff0ff0ff0ff0ff0ff0ff
(0) ff19 
kifu:
☗9三量子成 [0,[9,3,[1,"ff"],[0,"fg"]],[9,7,[0,"ff"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗8三成量子 [0,[8,3,[1,"ff"],[0,"fg"]],[9,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗8二成量子 [0,[8,2,[1,"ff"],[0,"fg"]],[8,3,[0,"fg"],[]],[1,0,"ff"]]
☖6二量子 [1,[6,2,[],[1,"ff"]],[6,1,[1,"ff"],[]]]
☗8一成量子 [0,[8,1,[1,"ff"],[0,"fg"]],[8,2,[0,"fg"],[]],[1,0,"ff"]]
☖7二量子 [1,[7,2,[],[1,"ff"]],[7,1,[1,"ff"],[]]]
☗同成量子 [0,[7,2,[1,"ff"],[0,"fg"]],[8,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗7三成量子 [0,[7,3,[1,"ff"],[0,"fg"]],[7,2,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6三成量子 [0,[6,3,[1,"ff"],[0,"fg"]],[7,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5三成量子 [0,[5,3,[1,"ff"],[0,"fg"]],[6,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4三成量子 [0,[4,3,[1,"ff"],[0,"fg"]],[5,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3三成量子 [0,[3,3,[1,"ff"],[0,"fg"]],[4,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2三成量子 [0,[2,3,[1,"ff"],[0,"fg"]],[3,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗1三成量子 [0,[1,3,[1,"ff"],[0,"fg"]],[2,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗1一成量子 [0,[1,1,[1,"ff"],[0,"fg"]],[1,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2二成量子 [0,[2,2,[1,"ff"],[0,"fg"]],[1,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2一成量子 [0,[2,1,[1,"ff"],[0,"fg"]],[2,2,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3一成量子 [0,[3,1,[1,"ff"],[0,"fg"]],[2,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4一成量子 [0,[4,1,[1,"ff"],[0,"fg"]],[3,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5一成量子 [0,[5,1,[1,"ff"],[0,"fg"]],[4,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6二成量子 [0,[6,2,[1,"ff"],[0,"fg"]],[5,1,[0,"fg"],[]],[1,0,"ff"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 108,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Quantum shogi can finish when everything is taken 2`] = `
ban:
(1) 
___________________________
1ff______0fg_______________
___________________________
___________________________
___________________________
___________________________
___0ff0ff0ff0ff0ff0ff0ff0ff
___0ff_______________0ff___
0ff0ff0ff0ff0ff0ff0ff0ff0ff
(0) ff19 
kifu:
☗9三量子成 [0,[9,3,[1,"ff"],[0,"fg"]],[9,7,[0,"ff"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗8三成量子 [0,[8,3,[1,"ff"],[0,"fg"]],[9,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗8二成量子 [0,[8,2,[1,"ff"],[0,"fg"]],[8,3,[0,"fg"],[]],[1,0,"ff"]]
☖6二量子 [1,[6,2,[],[1,"ff"]],[6,1,[1,"ff"],[]]]
☗8一成量子 [0,[8,1,[1,"ff"],[0,"fg"]],[8,2,[0,"fg"],[]],[1,0,"ff"]]
☖7二量子 [1,[7,2,[],[1,"ff"]],[7,1,[1,"ff"],[]]]
☗同成量子 [0,[7,2,[1,"ff"],[0,"fg"]],[8,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗7三成量子 [0,[7,3,[1,"ff"],[0,"fg"]],[7,2,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6三成量子 [0,[6,3,[1,"ff"],[0,"fg"]],[7,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5三成量子 [0,[5,3,[1,"ff"],[0,"fg"]],[6,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4三成量子 [0,[4,3,[1,"ff"],[0,"fg"]],[5,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3三成量子 [0,[3,3,[1,"ff"],[0,"fg"]],[4,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2三成量子 [0,[2,3,[1,"ff"],[0,"fg"]],[3,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗1三成量子 [0,[1,3,[1,"ff"],[0,"fg"]],[2,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗1一成量子 [0,[1,1,[1,"ff"],[0,"fg"]],[1,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2二成量子 [0,[2,2,[1,"ff"],[0,"fg"]],[1,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2一成量子 [0,[2,1,[1,"ff"],[0,"fg"]],[2,2,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3一成量子 [0,[3,1,[1,"ff"],[0,"fg"]],[2,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4一成量子 [0,[4,1,[1,"ff"],[0,"fg"]],[3,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5一成量子 [0,[5,1,[1,"ff"],[0,"fg"]],[4,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6二成量子 [0,[6,2,[1,"ff"],[0,"fg"]],[5,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 108,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Quantum shogi can finish when everything is taken 3`] = `
ban:
(1) 
___________________________
0fg________________________
___________________________
___________________________
___________________________
___________________________
___0ff0ff0ff0ff0ff0ff0ff0ff
___0ff_______________0ff___
0ff0ff0ff0ff0ff0ff0ff0ff0ff
(0) ff20 
kifu:
☗9三量子成 [0,[9,3,[1,"ff"],[0,"fg"]],[9,7,[0,"ff"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗8三成量子 [0,[8,3,[1,"ff"],[0,"fg"]],[9,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗8二成量子 [0,[8,2,[1,"ff"],[0,"fg"]],[8,3,[0,"fg"],[]],[1,0,"ff"]]
☖6二量子 [1,[6,2,[],[1,"ff"]],[6,1,[1,"ff"],[]]]
☗8一成量子 [0,[8,1,[1,"ff"],[0,"fg"]],[8,2,[0,"fg"],[]],[1,0,"ff"]]
☖7二量子 [1,[7,2,[],[1,"ff"]],[7,1,[1,"ff"],[]]]
☗同成量子 [0,[7,2,[1,"ff"],[0,"fg"]],[8,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗7三成量子 [0,[7,3,[1,"ff"],[0,"fg"]],[7,2,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6三成量子 [0,[6,3,[1,"ff"],[0,"fg"]],[7,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5三成量子 [0,[5,3,[1,"ff"],[0,"fg"]],[6,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4三成量子 [0,[4,3,[1,"ff"],[0,"fg"]],[5,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3三成量子 [0,[3,3,[1,"ff"],[0,"fg"]],[4,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2三成量子 [0,[2,3,[1,"ff"],[0,"fg"]],[3,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗1三成量子 [0,[1,3,[1,"ff"],[0,"fg"]],[2,3,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗1一成量子 [0,[1,1,[1,"ff"],[0,"fg"]],[1,3,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗2二成量子 [0,[2,2,[1,"ff"],[0,"fg"]],[1,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗2一成量子 [0,[2,1,[1,"ff"],[0,"fg"]],[2,2,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗3一成量子 [0,[3,1,[1,"ff"],[0,"fg"]],[2,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗4一成量子 [0,[4,1,[1,"ff"],[0,"fg"]],[3,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗5一成量子 [0,[5,1,[1,"ff"],[0,"fg"]],[4,1,[0,"fg"],[]],[1,0,"ff"]]
☖9一量子 [1,[9,1,[],[1,"ff"]],[9,2,[1,"ff"],[]]]
☗6二成量子 [0,[6,2,[1,"ff"],[0,"fg"]],[5,1,[0,"fg"],[]],[1,0,"ff"]]
☖9二量子 [1,[9,2,[],[1,"ff"]],[9,1,[1,"ff"],[]]]
☗同成量子 [0,[9,2,[1,"ff"],[0,"fg"]],[6,2,[0,"fg"],[]],[1,0,"ff"]]
☗勝ち "_1"
{
  "version": "0.0",
  "status": {
    "num": 2,
    "message": "王に確定した駒を取りました"
  },
  "ruleid": 108,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Quantum shogi can move 1`] = `
ban:
(1) 
1ff1ff1ff1ff1ff1ff1ff1ff1ff
___1ff_______________1ff___
1ff1ff1ff1ff1ff1ff1ff1ff1ff
___________________________
_________0ff_______________
___________________________
0ff0ff___0ff0ff0ff0ff0ff0ff
___0ff_______________0ff___
0ff0ff0ff0ff0ff0ff0ff0ff0ff
(0) 
kifu:
☗6五量子 [0,[6,5,[],[0,"ff"]],[7,7,[0,"ff"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 108,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Quantum shogi can move 2`] = `
Array [
  Object {
    "XY": (7, 3),
    "type": 1,
  },
  Object {
    "XY": (5, 3),
    "type": 1,
  },
]
`;

exports[`Quantum shogi can put piece with id 1`] = `
Array [
  Object {
    "XY": (8, 7),
    "type": 1,
  },
]
`;

exports[`Quantum shogi can put piece with id 2`] = `
Array [
  Object {
    "XY": (9, 6),
    "type": 2,
  },
  Object {
    "XY": (9, 7),
    "type": 2,
  },
  Object {
    "XY": (8, 5),
    "type": 2,
  },
  Object {
    "XY": (7, 5),
    "type": 2,
  },
  Object {
    "XY": (6, 5),
    "type": 2,
  },
  Object {
    "XY": (9, 4),
    "type": 2,
  },
  Object {
    "XY": (9, 3),
    "type": 2,
  },
  Object {
    "XY": (9, 6),
    "type": 2,
  },
  Object {
    "XY": (9, 7),
    "type": 2,
  },
]
`;
